src/Middleware/RequestDecompression/src/DeflateDecompressionProvider.cs

```diff
@@ -13,6 +14,221 @@ internal sealed class DeflateDecompressionProvider : IDecompressionProvider
     /// <inheritdoc />
     public Stream GetDecompressionStream(Stream stream)
     {
-        return new DeflateStream(stream, CompressionMode.Decompress, leaveOpen: true);
+        return new ZLibOrDeflateStream(stream);
+    }
+
+    // As described in RFC 2616, the deflate content-coding token represents the "zlib" format
+    // (RFC 1950) in combination with the "deflate" compression algorithm (RFC 1951). However,
+    // in practice, it is also possible for raw, unwrapped deflate compression to be used with
+    // "deflate" as the content-encoding. This class lets us wrap either a zlib- or deflate-
+    // compressed stream and delay figuring out what it is until the first read.
+    internal sealed class ZLibOrDeflateStream : Stream
+    {
+        private readonly PeekFirstByteReadStream _stream;
+        private Stream? _decompressionStream;
+
+        public ZLibOrDeflateStream(Stream stream) => _stream = new PeekFirstByteReadStream(stream);
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _decompressionStream?.Dispose();
+                _stream.Dispose();
+            }
+
+            base.Dispose(disposing);
+        }
+
+        public override bool CanRead => true;
```

Comment:
are these right? I think all the members of ZLibOrDeflateStream should operate directly on _decompressionStream (initializing first if needed) and not on _stream unless it's doing the peeking (nor hard coded either). This will allow PeekFirstByteReadStream to be almost all stubbed as only PeekFirstByteXXX() will ever be called on it.

Reply:
This stream is used for reading and only for reading, and it shouldn't be here by construction of the underlying stream isn't readable. Delegating would just add more virtual dispatch.